src/datalink/bpf.rs

```diff
@@ -139,30 +139,29 @@ pub fn datalink_channel(network_interface: &NetworkInterface,
     }
 
     let fd = Arc::new(internal::FileDesc { fd: fd });
-    let sender = DataLinkSenderImpl {
+    let sender = Box::new(DataLinkSenderBpf {
         fd: fd.clone(),
         write_buffer: repeat(0u8).take(write_buffer_size).collect(),
         header_size: header_size,
-    };
-    let receiver = DataLinkReceiverImpl {
+    });
+    let receiver = Box::new(DataLinkReceiverBpf {
         fd: fd,
         read_buffer: repeat(0u8).take(read_buffer_size).collect(),
         header_size: header_size,
-    };
+    });
 
     Ok((sender, receiver))
 }
 
-pub struct DataLinkSenderImpl {
+pub struct DataLinkSenderBpf {
     fd: Arc<internal::FileDesc>,
     write_buffer: Vec<u8>,
     header_size: usize,
 }
 
-impl DataLinkSenderImpl {
-    pub fn build_and_send<F>(&mut self, num_packets: usize, packet_size: usize,
-                          func: &mut F) -> Option<io::Result<()>>
-        where F : FnMut(MutableEthernetPacket)
+impl DataLinkSender for DataLinkSenderBpf {
+    fn build_and_send(&mut self, num_packets: usize, packet_size: usize,
+                          func: &mut FnMut(MutableEthernetPacket)) -> Option<io::Result<()>>
```

Comment:
This change worries me a lot - `build_and_send` now requires a dynamic dispatch, rather than static, which has the potential to kill performance - this is not so much an issue for the BPF backend, but for the likes of netmap, and potential DPDK or PF_RING ZC backends where it's possible to work with 10GbE links and every cycle counts this could be a killer.

The reason it was done the old way was to avoid this issue. Perhaps there is a happy middle ground we could find? (Or, alternatively, if you can give before/after assembly proving that the dynamic dispatch doesn't happen in practice, that would be fine too).